src/SynthesiserRelation.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2018, The Souffle Developers. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+#pragma once
+
+#include "Global.h"
+#include "IndexSetAnalysis.h"
+#include "RamRelation.h"
+
+#include <numeric>
+
+namespace souffle {
+
+class SynthesiserRelation {
+protected:
+    /** Ram relation referred to by this */
+    const RamRelation& relation;
+
+    /** Indices used for this relation */
+    const IndexSet& indices;
+
+    /** The data structure used for the relation */
+    std::string dataStructure;
+
+    /** The final list of indices used */
+    std::vector<std::vector<int>> computedIndices;
```

Comment:
Unless there's a specific reason for `int`, use something more specific like we do elsewhere. e.g. uint32_t or whatever is appropriate.